src/Microsoft.Extensions.WebEncoders.Core/UrlPathDecoder.cs

```diff
@@ -0,0 +1,341 @@
+                }
+                if (remainingBytes == 2 && currentDecodeBits >= 0x110)
+                {
+                    // This is going to be out of the upper Unicode bound 0x10FFFF.
+                    return 0;
+                }
+            }
+
+            if (currentDecodeBits < expectValueMin)
+            {
+                // Overlong encoding (e.g. using 2 bytes to encode something that only needed 1).
+                return 0;
+            }
+
+            try
+            {
+                // Convert the code point into char
+                count = _utf8.GetCharCount(bytesBuffer, 0, bytesCount);
+                if (count > output.Length)
+                {
+                    output = new char[count];
+                }
+
+                _utf8.GetChars(bytesBuffer, 0, bytesCount, output, 0);
+                return bytesCount * 3;
+            }
+            catch (DecoderFallbackException)
+            {
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Copy characters in source string to target char array. After copied the indices points at the beginning
+        /// offset of both source and destination are updated to point to the position after the last copied character
+        /// </summary>
+        private static int Copy(char[] source, ref int sourceStart, char[] destination, ref int destinatonStart, int length)
+        {
+            for (var i = 0; i < length; ++i)
+            {
+                destination[destinatonStart++] = source[sourceStart++];
+
+                if (destinatonStart >= destination.Length ||
+                    sourceStart >= source.Length)
+                {
+                    return i + 1;
+                }
+            }
+
+            return length;
+        }
+
+        /// <summary>
+        /// Find the next % in the sequence. If % is not found, return the sequence length.
+        /// </summary>
+        private static int GetNextEncoded(int start, char[] array)
+        {
+            for (var i = start; i < array.Length; ++i)
+            {
+                if (array[i] == '%')
+                {
+                    return i;
+                }
+            }
+
+            return array.Length;
+        }
+
+        private static bool TryGetUnescapedByte(char first, char second, out byte result)
+        {
+            if (!IsHex(first) || !IsHex(second))
+            {
+                result = default(byte);
+                return false;
+            }
+
+            // result in range [0000, 00FF]
+            result = (byte)((HexToDec(first) << 4) + (HexToDec(second)));
+            return true;
+        }
+
+        private static bool IsHex(char value)
+        {
+            return (((value >= '0') && (value <= '9')) ||
+                    ((value >= 'A') && (value <= 'F')) ||
+                    ((value >= 'a') && (value <= 'f')));
+        }
+
+        private static int HexToDec(char value)
```

Comment:
On another look I'll not change it. It is consumed by `TryGetUnescapedByte` only and return `int`  avoid casting in the body.